### PR TITLE
Adding Output Support for Double2DArray for Client Generator and Service and updating Sample Measurement Example

### PR DESCRIPTION
**Pull request by Blahpapto1**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

Adding Support for Double2DArray Output for Client Generator and Service and updating Sample Measurement Example.

### Why should this Pull Request be merged?

Add Double2DArray Output support for measurements.

### What testing has been done?

- All automated tests for Client Generator and Service are running fine after adding Double2DArray output parameter.
- Measurement is running fine through command-line.
- Additional testing on MeasurementUI Editor by dropping Control and Indicator and connecting to measurement produce desired behavior.

### CODE REVIEW

**Comment by dixonjoel** [comment]
The code changes look good. We need to make at least these changes before it's ready to go in.
1. Add tests. See https://github.com/search?q=repo%3Ani%2Fmeasurement-plugin-python%20doublexydata&type=code for examples of where DoubleXYData is tested. We should have at least those tested for Double2DArray as well.
2. Get array.proto checked into ni-apis, update the third_party commit to pull array.proto in from there and revert all array.proto changes here.

@bkeryan Please take a look at these preliminary changes to make sure there's no big problems you see.

**Comment by dixonjoel** [comment]
@Blahpapto1 #1022 will add array.proto stubs you can use.

File: examples/sample_measurement/measurement.py [hunk 1]
```diff
@@ -42,6 +43,11 @@ class Color(Enum):
 @measurement_service.configuration("Bool In", nims.DataType.Boolean, False)
 @measurement_service.configuration("String In", nims.DataType.String, "sample string")
 @measurement_service.configuration("Enum In", nims.DataType.Enum, Color.BLUE, enum_type=Color)
+@measurement_service.configuration(
```
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:46]: The client generator doesn't support message configurations: https://github.com/ni/measurement-plugin-python/blob/main/packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py#L126

File: examples/sample_measurement/measurement.py [hunk 2]
```diff
@@ -56,6 +62,7 @@ class Color(Enum):
 @measurement_service.output("Bool out", nims.DataType.Boolean)
 @measurement_service.output("String out", nims.DataType.String)
 @measurement_service.output("Enum out", nims.DataType.Enum, enum_type=Color)
+@measurement_service.output("Double2DArray out", nims.DataType.Double2DArray)
```
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:65]: The client generator doesn't support message outputs other than DoubleXYData yet: https://github.com/ni/measurement-plugin-python/blob/main/packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py#L156

File: examples/sample_measurement/measurement.py (hunk 1 above)
Comment (bkeryan) [reply examples/sample_measurement/measurement.py:46]: You should also update https://github.com/ni/measurement-plugin-python/blob/main/packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py in order to test this.

**Comment by bkeryan** [comment]
Please also update the test measurements and add relevant test cases.
- Client generator test measurement: https://github.com/ni/measurement-plugin-python/blob/main/packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py
- Service test measurement: https://github.com/ni/measurement-plugin-python/blob/main/packages/service/tests/utilities/measurements/loopback_measurement/__init__.py

File: examples/sample_measurement/measurement.py (hunk 2 above)
Comment (Blahpapto1) [reply examples/sample_measurement/measurement.py:60]: Added Support for output for Double2DArray in Client Generator in newer commits.

**Comment by Blahpapto1** [comment]
The Failure in [PR / Run CI / Check Examples](https://github.com/ni/measurement-plugin-python/actions/runs/12011815055/job/33481655684?pr=1020) is due to sample_measurement using the 2.1.0 released version of nims which doesn't contain definition for Double2DArray. Should I update pyproject.toml to use pre-release version for nims ?

**Comment by bkeryan** [comment]
> The Failure in [PR / Run CI / Check Examples](https://github.com/ni/measurement-plugin-python/actions/runs/12011815055/job/33481655684?pr=1020) is due to sample_measurement using the 2.1.0 released version of nims which doesn't contain definition for Double2DArray. Should I update pyproject.toml to use pre-release version for nims containing definition for Double2DArray ?

Yes. See https://github.com/ni/measurement-plugin-python/wiki/Release-Process#example-versioning

File: packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py [hunk 3]
```diff
@@ -86,6 +86,7 @@ def _create_client(
     output_parameters_with_type = get_output_parameters_with_type(
         output_metadata, built_in_import_modules, custom_import_modules, enum_values_by_type
     )
+    custom_import_modules.sort()
```
Comment (dixonjoel) [inline packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py:89]: Why is this necessary?

File: packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py [hunk 4]
```diff
@@ -294,6 +299,13 @@ def get_output_parameters_with_type(
             if metadata.repeated:
                 parameter_type = f"typing.Sequence[{parameter_type}]"
 
+        if metadata.message_type and metadata.message_type == "ni.protobuf.types.Double2DArray":
+            parameter_type = "Double2DArray"
+            custom_import_modules.append(_2DARRAY_DATA_IMPORT)
+
+            if metadata.repeated:
```
Comment (dixonjoel) [inline packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py:306]: We are not trying to support a repeated 2D array. I think this should throw an exception.

File: packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py [hunk 5]
```diff
@@ -134,6 +138,7 @@ def measure(
     enum_output = enum_input
     enum_array_output = enum_array_input
     protobuf_enum_output = protobuf_enum_input
+    double2d_array_output = array_pb2.Double2DArray()
```
Comment (dixonjoel) [inline packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py:141]: I'd like to see us return some sample data here so in the test we make sure all the data is coming through with its rows and columns.

**Review by dixonjoel — CHANGES_REQUESTED** [review]
Various changes requested

File: packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py [hunk 6]
```diff
@@ -153,9 +154,13 @@ def get_configuration_and_output_metadata_by_index(
 
     output_parameter_list = []
     for output in metadata.measurement_signature.outputs:
-        if output.message_type and output.message_type != "ni.protobuf.types.DoubleXYData":
+        if (
+            output.message_type
+            and output.message_type != "ni.protobuf.types.DoubleXYData"
+            and output.message_type != "ni.protobuf.types.Double2DArray"
+        ):
```
Comment (bkeryan) [inline packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py:161]:
```suggestion
        if (
            output.message_type
            and output.message_type not in ["ni.protobuf.types.DoubleXYData", "ni.protobuf.types.Double2DArray"]
        ):
```

File: packages/generator/tests/acceptance/test_non_streaming_measurement_client.py [hunk 7]
```diff
@@ -65,6 +65,7 @@ def test___measurement_plugin_client___measure___returns_output(
         enum_out=EnumInEnum.BLUE,
         enum_array_out=[EnumInEnum.RED, EnumInEnum.GREEN],
         protobuf_enum_out=ProtobufEnumInEnum.BLACK,
+        double2darray_out=None,
```
Comment (bkeryan) [inline packages/generator/tests/acceptance/test_non_streaming_measurement_client.py:68]: Please test this parameter with a value other than None.

File: examples/sample_measurement/measurement.py [hunk 8]
```diff
@@ -60,6 +61,7 @@ class Color(Enum):
     "Protobuf Enum out", nims.DataType.Enum, enum_type=color_pb2.ProtobufColor
 )
 @measurement_service.output("String Array out", nims.DataType.StringArray1D)
+@measurement_service.output("Double2DArray Out", nims.DataType.Double2DArray)
```
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:64]: "When in Rome, do as the Romans do." By which I mean: look at the existing code and follow the same patterns and conventions.

```suggestion
@measurement_service.output("Double 2D Array Out", nims.DataType.Double2DArray)
```

File: packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py [hunk 9]
```diff
@@ -89,6 +91,7 @@ class Color(Enum):
 @measurement_service.output(
     "Protobuf Enum out", nims.DataType.Enum, enum_type=color_pb2.ProtobufColor
 )
+@measurement_service.output("Double2DArray out", nims.DataType.Double2DArray)
```
Comment (bkeryan) [inline packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py:94]:
```suggestion
@measurement_service.output("Double 2D Array out", nims.DataType.Double2DArray)
```

File: packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py (hunk 5 above)
Comment (bkeryan) [reply packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py:141]: Why didn't we do this for DoubleXYData?

Comment (bkeryan) [inline packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py:141]:
```suggestion
    double_2d_array_output = array_pb2.Double2DArray()
```

File: packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py [hunk 10]
```diff
@@ -153,9 +154,13 @@ def get_configuration_and_output_metadata_by_index(
...
             raise click.ClickException(
-                f"Measurement outputs do not support {output.message_type}. Only DoubleXYData is supported."
+                f"Measurement outputs do not support {output.message_type}. Only DoubleXYData and Double2DArray is supported."
```
Comment (bkeryan) [inline packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py:163]:
```suggestion
                f"Measurement outputs do not support {output.message_type}. Only DoubleXYData and Double2DArray are supported."
```

File: examples/sample_measurement/measurement.py
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:1]: > What does this Pull Request accomplish?
> Adding Double2DArray in Sample Measurement example.

This PR does a lot more than that. Please update the PR description to say what it accomplishes.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py (hunk 3 above)
Comment (Blahpapto1) [reply packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py:89]: `get_output_parameters_with_type` adds custom import modules by what value it gets first from output_metadata.values() [here](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py#L279C5-L279C46) and since the DoubleXYData value is before Double2DArray value [here](https://github.com/ni/measurement-plugin-python/blob/7ea52e5a2b45da2af95e3f66e0563a41239ccfdd/packages/generator/tests/test_assets/example_renders/measurement_plugin_client/non_streaming_data_measurement_client.py#L376C13-L385C15) it is picking that up first. But then it throws an error later down the line regarding import modules not being in correct order.

Now, on retrospecting it would be better to just add Double2DArray before DoubleXYData but this needs to be addressed that in these types of cases we need to be mindful of the import order

File: packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py (hunk 4 above)
Comment (Blahpapto1) [reply packages/generator/ni_measurement_plugin_sdk_generator/client/_support.py:305]: Updated in the latest commit to raise exception if repeated is True

File: examples/sample_measurement/measurement.py [hunk 11]
```diff
@@ -85,7 +94,8 @@ def cancel_callback() -> None:
     string_output = string_input
     enum_output = enum_input
     protobuf_enum_output = protobuf_enum_input
-    string_array_out = string_array_in
+    string_array_output = string_array_in
+    double2d_array_output = array_pb2.Double2DArray()
```
[the remark this answers is not in the document]
Comment (Blahpapto1) [reply examples/sample_measurement/measurement.py]: I am handling this in the Input Support PR. All these empty value Double2DArray instances and None value tests are replaced with some default value in the other PR.

Will also update the MeasUI for Sample Measurement as part of Input Support PR.

File: packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py (hunk 5 above)
Comment (Blahpapto1) [reply packages/generator/tests/utilities/measurements/non_streaming_data_measurement/__init__.py]: Refer to this [comment](https://github.com/ni/measurement-plugin-python/pull/1020#discussion_r1871044195)

File: packages/generator/tests/acceptance/test_non_streaming_measurement_client.py (hunk 7 above)
Comment (Blahpapto1) [reply packages/generator/tests/acceptance/test_non_streaming_measurement_client.py]: Refer to this [comment](https://github.com/ni/measurement-plugin-python/pull/1020#discussion_r1871044195)

File: packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py (hunk 3 above)
Comment (Blahpapto1) [reply packages/generator/ni_measurement_plugin_sdk_generator/client/__init__.py:89]: Value of custom_import_module if we are not sorting - 
['from ni_measurement_plugin_sdk_service._internal.stubs.ni.protobuf.types.xydata_pb2 import DoubleXYData', 'from ni_measurement_plugin_sdk_service._internal.stubs.ni.protobuf.types.array_pb2 import Double2DArray']

File: examples/sample_measurement/measurement.py
Comment (Blahpapto1) [reply examples/sample_measurement/measurement.py:1]: Updated description.

**Review by bkeryan — APPROVED** [review]

**Review by dixonjoel — APPROVED** [review]
